logutils/namespaces_tree.go

```diff
@@ -0,0 +1,137 @@
+package logutils
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+	"sync"
+
+	"go.uber.org/zap/zapcore"
+)
+
+var namespacesRegex = regexp.MustCompile(`^([a-zA-Z0-9_\.]+:(debug|info|warn|error))(,[a-zA-Z0-9_\.]+:(debug|info|warn|error))*$`)
+var errInvalidNamespacesFormat = fmt.Errorf("invalid namespaces format")
+
+type namespacesTree struct {
+	namespaces map[string]*namespaceNode
+	minLevel   zapcore.Level
+	mu         sync.RWMutex
+}
+
+type namespaceNode struct {
+	level    zapcore.Level
+	children map[string]*namespaceNode
+}
+
+func newNamespacesTree() *namespacesTree {
+	return &namespacesTree{
+		namespaces: make(map[string]*namespaceNode),
+		minLevel:   zapcore.InvalidLevel,
+	}
+}
+
+// buildNamespacesTree creates a new tree instance from a string of namespaces.
+// The namespaces string should be a comma-separated list of namespace:level pairs,
+// where level is one of: "debug", "info", "warn", "error".
+//
+// Example: "namespace1:debug,namespace2.namespace3:error"
+func buildNamespacesTree(namespaces string) (*namespacesTree, error) {
+	tree := newNamespacesTree()
+
+	err := tree.Rebuild(namespaces)
+	if err != nil {
+		return nil, err
+	}
+
+	return tree, nil
+}
+
+// LevelFor finds the log level for the given namespace by searching for the closest match in the tree.
+// In case of no match, it returns zapcore.InvalidLevel.
+func (t *namespacesTree) LevelFor(namespace string) zapcore.Level {
+	t.mu.RLock()
+	defer t.mu.RUnlock()
+
+	currentNode := t.namespaces
+	var lastFound *namespaceNode
+
+	for _, name := range strings.Split(namespace, ".") {
+		if node, exists := currentNode[name]; exists {
+			lastFound = node
+			currentNode = node.children
+		} else {
+			break
+		}
+	}
+
+	if lastFound != nil {
+		return lastFound.level
+	}
+
+	// Default to zapcore.InvalidLevel if no match found.
+	return zapcore.InvalidLevel
+}
+
+// Rebuild updates the tree with a new set of namespaces.
+func (t *namespacesTree) Rebuild(namespaces string) error {
+	if namespaces == "" {
+		t.mu.Lock()
+		defer t.mu.Unlock()
+		t.namespaces = make(map[string]*namespaceNode)
+		t.minLevel = zapcore.InvalidLevel
+		return nil
+	}
+
+	if !namespacesRegex.MatchString(namespaces) {
+		return errInvalidNamespacesFormat
+	}
+
+	t.mu.Lock()
+	defer t.mu.Unlock()
```

Comment:
I guess for code simplicity it can be done once in the beginning of the function? There's no need not to lock the mutex for the `namespacesRegex.MatchString`, right?